### PR TITLE
Keyboard shortcuts

### PR DESCRIPTION
**Pull request by spatializes**
#### What's this PR do?

Adds the following keyboard shortcuts (selections with **_bold**_ are not implemented, my comments in [ ] when needed):
- Ctrl-A - Select all text in current source code
- Ctrl-C - Copy selected text to the clipboard
- Ctrl-F - Find or replace text
- Ctrl-V - Paste text from clipboard to selected area
- Ctrl-X - Cut selected text to the clipboard
- Ctrl-Z - Undo last action (unlimited)[200 by default]
- Ctrl-Shift-Z - Redo last action (unlimited)
- Tab - Indent block (Inserts space characters)
- Shift-Tab - Outdent block (Deletes space characters)
- F3 - Find text again
- Shift+F3 - Find text again (opposite direction)
- **_F4 - Replace current found selection**_
- **_Ctrl-F4 - Perform replace and find next**_
- **_Ctrl-Home - Jump to top of file**_
- **_Ctrl-End - Jump to end of file**_
- Ctrl-CursorUp - Move source view up one line without moving cursor [moves by distance of a line]
- Ctrl-CursorDown - Move source view down one line without moving cursor [moves by distance of a line]
- **_Ctrl-+/- (or =/_) - increase/decrease font size**_
- Ctrl-O - Open source; may just focus file explorer
- **_Cursor + enter keys navigate and select files**_
- **_Backspace exits file view (project) (to project view)**_
- Ctrl-S - Save current source code file to its current location on disk
- Ctrl-P - Print current source [this currently prints the whole page. Needs to be updated to only print the code area]
- Ctrl-Tab - Switch to next open source [ctrl + tab does not register a keydown event, so I added some logic to Iggins to case off this scenario]
- Ctrl-Shift-Tab - Switch to previous open source [same issue as ctrl + tab]
- **_F6 or Ctrl-I - Identify BASIC Stamp firmware.**_
- **_F7 or Ctrl-T - Perform a syntax check on the code.**_
- **_F8 or Ctrl-M - Tokenize and open Memory Map display.**_
- **_F9 or Ctrl-R - Tokenize, download to BASIC Stamp, and open Debug display if necessary.**_
- **_ESC - Close current dialog.**_

This PR exposes methods to Iggins to be called by keyboard shortcuts:
- open file
- save file

This PR also adds functionality that was needed to enable these shortcuts, and exposes them to iggins:
- move to next file
- move to previous file
- scroll one line at a time
#### Where should the reviewer start?

Load this branch onto the Chromebook. Also, be sure Iggins in package.json is pointing to the branch 'Rough-Iggin-Core' from the Iggins library. Reload the extension. Create a few test files. Try out the shortcuts that are enabled above.
#### How should this be manually tested?

Try to press each of the shortcuts above after loading in this branch.  They will also run on a mac, however some of the functionality built into CodeMirror uses the "Command" key, such as CMD + A, CMD + C, CMD + V.
#### Any background context you want to provide?

I first implemented these shortcuts using CodeMirror's built in keypress system. See previous commits for that setup. I then refactored everything to use Iggins instead.
#### What are the relevant tickets?

Working on issue #88
#### Screenshots (if appropriate)

NA
#### Questions:
- Is there a blog post?
  No
- Does the knowledge base need an update?
  Yes

### CODE REVIEW

File: plugins/editor/key-extension.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      },
+      moveByScrollUpLine: function() {
+        const scrollbox = cm.getScrollInfo();
+        cm.scrollTo(null, scrollbox.top - cm.defaultTextHeight());
+      },
+      moveByScrollDownLine: function() {
+        const scrollbox = cm.getScrollInfo();
+        cm.scrollTo(null, scrollbox.top + cm.defaultTextHeight());
+      },
+      Tab: function() {
```
Comment (phated) [inline plugins/editor/key-extension.js]: unused

File: plugins/sidebar/file-list.js [hunk 2]
```diff
@@ -4,6 +4,36 @@ const React = require('react');
...
+  componentDidMount: function(){
+    app.keypress.register('CTRL_TAB', this.previousFile);
+    app.keypress.register('CTRL_SHIFT_TAB', this.nextFile);
+  },
+  componentWillUnmount: function(){
+    app.keypress.register('CTRL_TAB');
+    app.keypress.register('CTRL_SHIFT_TAB');
+  },
+  previousFile: function(){
+    this.changeFile(true);
```
Comment (phated) [inline plugins/sidebar/file-list.js]: boolean arguments are bad, this should be an object with `direction` property.  e.g. `{ direction: 'next' }` `{ direction: 'prev' }` or something similar

File: plugins/sidebar/file-list.js [hunk 3]
```diff
@@ -4,6 +4,36 @@ const React = require('react');
...
+  },
+  nextFile: function() {
+    this.changeFile(false);
```
Comment (phated) [inline plugins/sidebar/file-list.js]: same as above

File: plugins/sidebar/file-list.js [hunk 4]
```diff
@@ -4,6 +4,36 @@ const React = require('react');
...
+    const space = this.props.workspace;
+    const filename = space.filename.deref();
+
+    space.directory.forEach(function(x, i) {
+      if(x.get('name') === filename) {
+        if(i === space.directory.size - 1) {
+          i = -1;
+        }
+        const shift = moveup ? i - 1 : i + 1;
+        const switchFile = space.directory.get(shift).get('name');
```
Comment (phated) [inline plugins/sidebar/file-list.js]: this can use `getIn([shift, 'name'])`

File: plugins/sidebar/file-list.js [hunk 5]
```diff
@@ -4,6 +4,36 @@ const React = require('react');
 const List = require('react-material/components/List');
 
 const FileList = React.createClass({
+
+  componentDidMount: function(){
+    app.keypress.register('CTRL_TAB', this.previousFile);
+    app.keypress.register('CTRL_SHIFT_TAB', this.nextFile);
+  },
+  componentWillUnmount: function(){
+    app.keypress.register('CTRL_TAB');
```
Comment (phated) [inline plugins/sidebar/file-list.js]: why does `register` unregister? bad API

File: plugins/editor/key-extension.js (hunk 1 above)
Comment (spatializes) [inline plugins/editor/key-extension.js]: these are being used to tap into CodeMirror command functionality. 

Ctrl + up should use CodeMirror's command API to get the current scroll location and move up by whatever the CodeMirror default text height is.

after searching with Ctrl + f (CodeMirror built in search function), this maps F3 and Shit + F3 to find next or find previous.

File: plugins/sidebar/file-list.js (hunk 5 above)
Comment (spatializes) [inline plugins/sidebar/file-list.js]: should be unregister, updating.

File: plugins/editor/key-extension.js (hunk 1 above)
Comment (phated) [inline plugins/editor/key-extension.js]: I meant that the `Tab` function isn't assigned to any key presses

File: plugins/sidebar/file-list.js (hunk 4 above)
Comment (spatializes) [inline plugins/sidebar/file-list.js]: updated with getIn()

File: plugins/sidebar/file-list.js (hunk 2 above)
Comment (spatializes) [inline plugins/sidebar/file-list.js]: fixed this

File: plugins/editor/key-extension.js (hunk 1 above)
Comment (spatializes) [inline plugins/editor/key-extension.js]: oh right, did not see that. added the register for tab.

**Comment by phated** [comment]
Needs print and esc to close modals

**Comment by spatializes** [comment]
Updated with latest fixes.

**Pull request merged**